tests/test_ledger.py

```diff
@@ -1,445 +1,147 @@
-
-    # Mirror with duplicate transactions and delete=False
-    target = pd.concat(
-        [
-            LEDGER_ENTRIES.query("id == 1"),
-            LEDGER_ENTRIES.query("id == 1").assign(id=5),
-            LEDGER_ENTRIES.query("id == 2").assign(id=6),
-            LEDGER_ENTRIES.query("id == 2"),
+    def test_add_ledger_entry(self, ledger, ledger_id):
+        ledger.restore(accounts=self.ACCOUNTS, vat_codes=self.VAT_CODES)
+        target = self.LEDGER_ENTRIES.query("id == @ledger_id")
+        id = ledger.add_ledger_entry(target)
+        remote = ledger.ledger()
+        created = remote.loc[remote["id"] == str(id)]
+        expected = ledger.standardize_ledger(target)
+        assert_frame_equal(
+            created, expected, ignore_index=True, ignore_columns=["id"], check_exact=True
+        )
+
+    def test_add_ledger_with_non_existing_vat(self, ledger):
+        # Adding a ledger entry with non existing VAT code should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["vat_code"].iat[0] = "Test_Non_Existent_VAT_code"
+        with pytest.raises(ValueError, match="No id found for tax code"):
+            ledger.add_ledger_entry(target)
+
+    def test_add_ledger_with_non_existing_account(self, ledger):
+        # Adding a ledger entry with non existing account should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["account"] = 33333
+        with pytest.raises(ValueError, match="No id found for account"):
+            ledger.add_ledger_entry(target)
+
+    def test_add_ledger_with_non_existing_currency(self, ledger):
+        # Adding a ledger entry with non existing currency code should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["currency"] = "Non_Existent_Currency"
+        with pytest.raises(ValueError, match="No id found for currency"):
+            ledger.add_ledger_entry(target)
+
+    @pytest.mark.parametrize("id", [15, 16])
+    def test_adding_transaction_with_two_non_base_currencies_fails(self, ledger, id):
+        target = self.LEDGER_ENTRIES[self.LEDGER_ENTRIES["id"] == id]
+        expected = (
+            "CashCtrl allows only the base currency plus a single foreign currency"
+        )
+        with pytest.raises(ValueError, match=expected):
+            ledger.add_ledger_entry(target)
+
+    def test_modify_non_existed_raises_error(self, ledger):
+        super().test_modify_non_existed_raises_error(
+            ledger, error_class=RequestException, error_message="entry does not exist"
+        )
+
+    def test_update_ledger_with_illegal_attributes(self, ledger):
+        id = ledger.add_ledger_entry(self.LEDGER_ENTRIES.query("id == 1"))
+
+        # Updating a ledger with non existent VAT code should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["id"] = id
+        target["vat_code"] = "Test_Non_Existent_VAT_code"
```

Comment:
Again, I see a benefit in hard-coding the test data right here (in a concise csv or similar format).

(Applies throughout the method)